File: dynamic_programming/coin_change/coin_change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [100000] * (amount + 1)
        dp[0] = 0
        for i in range(1, amount + 1):
            for c in coins:
                if c <= i:
                    dp[i] = min(dp[i], 1 + dp[i - c])
        return dp[amount] if dp[amount] != 100000 else -1

File: dynamic_programming/coin_change/coin_change.java
class Solution {
  public int coinChange(int[] coins, int amount) {
    int[] dp = new int[amount + 1];
    for (int i = 1; i <= amount; i++) {
      int minNum = Integer.MAX_VALUE;
      for (int c : coins) {
        if (i - c >= 0 && dp[i - c] >= 0) {
          minNum = Math.min(minNum, dp[i - c]);
        }
      }
      if (minNum == Integer.MAX_VALUE) {
        dp[i] = -1;
      } else {
        dp[i] = minNum + 1;
      }
    }
    return dp[amount];
  }
}

class Test {
  public static void main(String[] args) {
    Solution s = new Solution();
    System.out.println(s.coinChange(new int[]{1, 2, 5}, 11));
    System.out.println(s.coinChange(new int[]{2}, 3));
    System.out.println(s.coinChange(new int[]{1}, 0));
  }
}